pkg/core/pipeline/scm/config.go

```diff
@@ -0,0 +1,154 @@
+	"strings"
+
+	"github.com/goware/urlx"
+	jschema "github.com/invopop/jsonschema"
+	"github.com/sirupsen/logrus"
+	"github.com/updatecli/updatecli/pkg/core/jsonschema"
+	"github.com/updatecli/updatecli/pkg/plugins/scms/git"
+	"github.com/updatecli/updatecli/pkg/plugins/scms/github"
+	"github.com/updatecli/updatecli/pkg/plugins/utils/gitgeneric"
+)
+
+type ConfigSpec interface {
+	Merge(interface{}) error
+}
+
+type Config struct {
+	Kind string
+	Spec interface{} `jsonschema:"type=object"`
+	// Disabled is a setting used to disable the local git repository auto configuration
+	Disabled bool
+}
+
+// Validate returns nil if the Config object is valid
+// or a wrapped error of all the failed validations otherwise
+func (c Config) Validate() error {
+	validationErrs := []string{}
+
+	if c.Disabled {
+		if c.Kind != "" {
+			validationErrs = append(validationErrs, "specified value for 'kind' found while SCM is disabled")
+		}
+		if c.Spec != nil {
+			validationErrs = append(validationErrs, "specified value for 'spec' found while SCM is disabled")
+		}
+	} else {
+		if c.Kind == "" {
+			validationErrs = append(validationErrs, "missing 'kind' value")
+		}
+		if c.Spec == nil {
+			validationErrs = append(validationErrs, "missing 'spec' value")
+		}
+	}
+
+	if len(validationErrs) > 0 {
+		return fmt.Errorf("%w: %s", ErrWrongConfig, strings.Join(validationErrs, ","))
+	}
+	return nil
+}
+
+// AutoGuess tries to fill the Config object receiver with the configuration auto-guessed from the provided directory
+// It only returns an error if it fails "hard" (can't guess SCM, etc.).
+func (c *Config) AutoGuess(configName, workingDir string, gitHandler gitgeneric.GitHandler) error {
+	remotes, err := gitHandler.RemoteURLs(workingDir)
+	if err != nil {
+		return err
+	}
+	logrus.Debugf("Found the following remotes in %s: %s", workingDir, remotes)
+
+	originRemoteURL, ok := remotes["origin"]
+	if !ok {
+		return fmt.Errorf("no remote named 'origin' could be found in the repository %s", workingDir)
+	}
+
+	var remoteHostname string
+	httpRemoteUrl, err := url.Parse(originRemoteURL)
+	if err != nil {
+		/**
+		Case of non HTTPS scheme: either git or SSH (github.com for instance).
+		Hostname extraction on the "left" of the colon character (':')
+		**/
+		leftUrl := strings.Split(originRemoteURL, ":")[0]
+		ux, _ := urlx.Parse(leftUrl)
+		remoteHostname, _, _ = urlx.SplitHostPort(ux)
+
+	} else {
+		remoteHostname = httpRemoteUrl.Hostname()
+	}
+
+	switch remoteHostname {
+	case "github.com":
+		slashEverywhereUrl := strings.Split(strings.ReplaceAll(originRemoteURL, ":", "/"), "/")
+		if len(slashEverywhereUrl) < 3 {
+			return fmt.Errorf("unable to parse the following GitHub remote URL: %s", slashEverywhereUrl)
+		}
+
+		autoguessSpec := github.Spec{
+			Repository: strings.TrimSuffix(slashEverywhereUrl[len(slashEverywhereUrl)-1], ".git"),
+			Owner:      slashEverywhereUrl[len(slashEverywhereUrl)-2],
+			Directory:  workingDir,
+			Branch:     "main",
```

Comment:
We should track that we need to automatically detech the branch as well in a separated PR/issue

Reply:
Sounds good to me, we can have smaller PR later.